lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -891,6 +891,93 @@ OpFoldResult OrOp::fold(ArrayRef<Attribute> constants) {
       constants, [](APInt &a, const APInt &b) { a |= b; });
 }
 
+/// Merge concats in an or op when they have non-conflicting fields.
+static bool canonicalizeOrExclusiveConcat(OrOp op, PatternRewriter &rewriter) {
+  auto inputs = op.inputs();
+  auto size = inputs.size();
+  auto concat1 = inputs.back().getDefiningOp<ConcatOp>();
+  auto concat2 = inputs[size - 2].getDefiningOp<ConcatOp>();
+
+  if (!concat1 || !concat2)
+    return false;
+
+  auto knownBits1 = computeKnownBits(inputs.back());
+  auto knownBits2 = computeKnownBits(inputs[size - 2]);
+  auto unknownBits1 = ~(knownBits1.Zero | knownBits1.One);
+  auto unknownBits2 = ~(knownBits2.Zero | knownBits2.One);
+  bool anyUnknownBitsOverlap = !(unknownBits1 & unknownBits2).isNullValue();
+  bool anyConstantOne = !(knownBits1.One | knownBits2.One).isNullValue();
+
+  // We can simplify only if unknown bits do not overlap. We don't handle the
+  // case where a constant 1 is present anywhere yet.
+  if (anyUnknownBitsOverlap || anyConstantOne)
+    return false;
+
+  SmallVector<Value> newConcatOperands;
+
+  // Simultaneously iterate over the operands of both concat ops, from MSB to
+  // LSB, pushing ranges of bits out at a time. Note that when we only use up
+  // part of a constant value, we mutate the operands array with a truncated
+  // version of the constant.
+  SmallVector<Value> operands1 = concat1->getOperands();
```

Comment:
don't need to make  copy of the operands just to iterate.

Reply:
Actually, I had to make a copy in order for my `*it1 = ...` assignments to work in the loop iterations when I truncate the constants. It probably isn't very kosher of me to copy the entire array just so that I can use it as temporary storage for the truncated values, but I couldn't think of a cleaner way to deal with it.

Would it better for me to just declare a variable outside the loop to hold the current operand? Something like:

```c++
auto it1 = operands1.begin();
auto operand1 = *it1;
++it1;

while (it1 != end1 && it2 != end2) {
  ...

  // Case where I need to truncate an operand
  operand1 = rewriter.create<hw::ConstantOp>(..., cst1.getValue().trunc(...));

  // Case where I want to advance the iterator;
  ++it1;
  operand1 = *it1;

  ...
}
```

Reply:
I tried doing this in https://github.com/richardxia/circt/commit/e0d9f74e14fddd4d6130bf783ddbb255c7c70848, but I have not pushed up to this branch since it ended up being very messy with having to constantly check whether it was safe to dereference the iterator pointer.

Is there a better way for me to iterate over the original operands while preserving the ability to maintain a bit of local state for the current amount of truncation of the constants?

Reply:
This has been addressed in my latest commit, where I used @lattner's approach of extracting subranges of the operands, where I no longer need to mutate the operands since I can now treat ConstantOps and other ops uniformly.